eyelid-match-ops/src/primitives/yashe.rs

```diff
@@ -104,14 +118,71 @@ impl<C: PolyConf> Yashe<C> {
         (priv_key, pub_key)
     }
 
+    pub fn encrypt(&self, m: Message<C>, public_key: PublicKey<C>, rng: &mut ThreadRng) -> Ciphertext<C> {
+        let logt = self.params.t.ilog2();
+        let s = self.sample_err(rng);
+        let e = self.sample_err(rng);
+        let mut c = s * public_key.h + e;
+        // TODO: find a way to access q and divide by t
+        let mut qdt: BigInt<2> = BigInt::from(Coeff::MODULUS);
+        //let qbt = q / Coeff::from(self.params.t);
+        qdt.divn(logt);
+        let mqdt = m.m * Coeff::from(qdt);
+        c = c + mqdt;
+        Ciphertext { c }
+    }
+
+    pub fn decrypt(&self, c: Ciphertext<C>, private_key: PrivateKey<C>) -> Message<C> {
+        let logt = self.params.t.ilog2();
+        let q = Coeff::MODULUS;
+        let qm1d2 = Coeff::MODULUS_MINUS_ONE_DIV_TWO;
+        let mut res = c.c * private_key.f;
+        for i in 0..C::MAX_POLY_DEGREE {
+            // Convert coefficient to big integer
+            let mut coeff_res = res[i].into_bigint();
+            // Multiply by t
+            coeff_res.muln(logt);
+            if coeff_res >= BigInt::from(0) {
+                // TODO: deal with carry
+                coeff_res.add_with_carry(&qm1d2);
+                // TODO: how to divide 2 bigints?
+                coeff_res = coeff_res / q;
+            } else {
+                // TODO: deal with borrow
+                coeff_res.sub_with_borrow(&qm1d2);
+                // TODO: how to divide 2 bigints?
+                coeff_res = coeff_res / q;
+            }
+            // TODO: validate the following
+            // REMARK: since t is a power of 2, we have that coeff_res.0[i] is zero for i > 0
+            // then we can apply the following formula
+            let result = coeff_res.0[0] % self.params.t;
+            res[i] = result.into();
+        }
```

Comment:
If you implement `DivAssign` as multiplying by `Coeff::inverse()`, then you can replace this entire loop with:
```suggestion
res /= C::t_as_coeff();
```

Just copy these impls, then make the calculation:
```rust
*coeff *= rhs.inverse().expect("can't divide by the zero coefficient");
```

https://github.com/Inversed-Tech/eyelid/blob/3668934d68780513ea61ede8f4ccfb2d6a7eaedb/eyelid-match-ops/src/primitives/poly/modular_poly/mul.rs#L17-L31

Reply:
We need division over the integers, instead of over the finite field. Then `inverse` is not going to help us here, since every element in the field has an inverse. What we need is division with remainder, but we are interested only in the quotient (and not in the remainder).